src/app/clusters/time-synchronization-server/time-synchronization-server.cpp

```diff
@@ -0,0 +1,980 @@
+    timeZoneOffset       = static_cast<int64_t>(tzStore.timeZone.offset);
+    VerifyOrReturnError(GetDSTOffset().size() != 0, CHIP_ERROR_INVALID_TIME);
+    const auto & dst = GetDSTOffset()[0];
+    dstOffset        = static_cast<int64_t>(dst.offset);
+
+    uint64_t usRemainder = chipEpochTime % chip::kMicrosecondsPerSecond;   // microseconds part of chipEpochTime
+    chipEpochTime        = (chipEpochTime / chip::kMicrosecondsPerSecond); // make it safe to cast to int64 by converting to seconds
+
+    uint64_t localTimeSec = static_cast<uint64_t>(static_cast<int64_t>(chipEpochTime) + timeZoneOffset + dstOffset);
+    localTime.SetNonNull((localTimeSec * chip::kMicrosecondsPerSecond) + usRemainder);
+    return CHIP_NO_ERROR;
+}
+
+TimeState TimeSynchronizationServer::GetUpdatedTimeZoneState()
+{
+    System::Clock::Microseconds64 utcTime;
+    auto tzList          = GetTimeZone();
+    size_t activeTzIndex = 0;
+    uint64_t chipEpochTime;
+
+    VerifyOrReturnValue(System::SystemClock().GetClock_RealTime(utcTime) == CHIP_NO_ERROR, TimeState::kInvalid);
+    VerifyOrReturnValue(tzList.size() != 0, TimeState::kInvalid);
+    VerifyOrReturnValue(UnixEpochToChipEpochMicro(utcTime.count(), chipEpochTime), TimeState::kInvalid);
+
+    for (size_t i = 0; i < tzList.size(); i++)
+    {
+        auto & tz = tzList[i].timeZone;
+        if (tz.validAt != 0 && tz.validAt <= chipEpochTime)
+        {
+            tz.validAt    = 0;
+            activeTzIndex = i;
+        }
+    }
+    if (activeTzIndex != 0)
+    {
+        mTimeZoneObj.size    = tzList.size() - activeTzIndex;
+        auto newTimeZoneList = tzList.SubSpan(activeTzIndex);
+        VerifyOrReturnValue(mTimeSyncDataProvider.StoreTimeZone(newTimeZoneList) == CHIP_NO_ERROR, TimeState::kInvalid);
+        VerifyOrReturnValue(LoadTimeZone() == CHIP_NO_ERROR, TimeState::kInvalid);
+        return TimeState::kChanged;
+    }
+    return TimeState::kActive;
+}
+
+TimeState TimeSynchronizationServer::GetUpdatedDSTOffsetState()
+{
+    System::Clock::Microseconds64 utcTime;
+    size_t activeDstIndex = 0;
+    auto dstList          = GetDSTOffset();
+    uint64_t chipEpochTime;
+    bool dstStopped = true;
+
+    VerifyOrReturnValue(System::SystemClock().GetClock_RealTime(utcTime) == CHIP_NO_ERROR, TimeState::kInvalid);
+    VerifyOrReturnValue(dstList.size() != 0, TimeState::kInvalid);
+    VerifyOrReturnValue(UnixEpochToChipEpochMicro(utcTime.count(), chipEpochTime), TimeState::kInvalid);
+
+    for (size_t i = 0; i < dstList.size(); i++)
+    {
+        if (dstList[i].validStarting <= chipEpochTime)
+        {
+            activeDstIndex = i;
+            dstStopped     = false;
+        }
+    }
+    VerifyOrReturnValue(dstStopped, TimeState::kStopped);
+    // if offset is zero and validUntil is null then no DST is used
+    if (dstList[activeDstIndex].offset == 0 && dstList[activeDstIndex].validUntil.IsNull())
+    {
+        return TimeState::kStopped;
+    }
```

Comment:
But in this case we don't drop the earlier entries from the list?  Followup to think about refactoring this code so we do that.

Reply:
Tracked in https://github.com/project-chip/connectedhomeip/issues/27575, resolving this comment for now, further followup on the issue.